src/olympia/blocklist/tests/test_mlbf.py

```diff
@@ -452,6 +452,145 @@ def test_diff_block_soft_to_hard(self):
+        covering all possible movements and the final diff state.
+
+        1. Not blocked -> Soft blocked
+        2. Not blocked -> Blocked
+        3. Soft blocked -> Blocked
+        4. Soft blocked -> Not blocked
+        5. Blocked -> Soft blocked
+        6. Blocked -> Not blocked
+
+        first stash
+        {
+          'blocked': [5,6],
+          'softblocked': [3,4],
+          # 1,2 are already unblocked so wouldn't be on the stash
+          'unblocked': [],
+        }
+
+        In old clients:
+        1: not blocked
+        2: not blocked
+        3: not blocked
+        4: not blocked
+        5: blocked
+        6: blocked
+
+        In new clients:
+        1: not blocked
+        2: not blocked
+        3: soft blocked
+        4: soft blocked
+        5: blocked
+        6: blocked
+
+        second stash
+        {
+          'blocked': [2,3],
+          'softblocked': [1,5],
+          'unblocked': [4,5,6],
+        }
+
+        In old clients:
+        1: not blocked
+        2: blocked
+        3: blocked
+        4: not blocked
+        5: not blocked
+        6: not blocked
+
+        In new clients:
+        1: soft blocked
+        2: blocked
+        3: blocked
+        4: not blocked
+        5: soft blocked
+        6: not blocked
+        """
+        # 1. Not blocked -> Soft blocked
+        one = addon_factory(guid='1', file_kw={'is_signed': True})
+        (one_hash,) = MLBF.hash_filter_inputs([(one.guid, one.current_version.version)])
+        # 2. Not blocked -> Blocked
+        two = addon_factory(guid='2', file_kw={'is_signed': True})
+        (two_hash,) = MLBF.hash_filter_inputs([(two.guid, two.current_version.version)])
+        # 3. Soft blocked -> Blocked
+        three, three_block = self._blocked_addon(
+            guid='3', block_type=BlockType.SOFT_BLOCKED, file_kw={'is_signed': True}
+        )
+        (three_hash,) = MLBF.hash_filter_inputs(
+            [(three.guid, three.current_version.version)]
+        )
+        # 4. Soft blocked -> Not blocked
+        four, four_block = self._blocked_addon(
+            guid='4', block_type=BlockType.SOFT_BLOCKED, file_kw={'is_signed': True}
+        )
+        (four_hash,) = MLBF.hash_filter_inputs(
+            [(four.guid, four.current_version.version)]
+        )
+        # 5. Blocked -> Soft blocked
+        five, five_block = self._blocked_addon(
+            guid='5', block_type=BlockType.BLOCKED, file_kw={'is_signed': True}
+        )
+        (five_hash,) = MLBF.hash_filter_inputs(
+            [(five.guid, five.current_version.version)]
+        )
+        # 6. Blocked -> Not blocked
+        six, six_block = self._blocked_addon(
+            guid='6', block_type=BlockType.BLOCKED, file_kw={'is_signed': True}
+        )
+        (six_hash,) = MLBF.hash_filter_inputs([(six.guid, six.current_version.version)])
+
+        first_mlbf = MLBF.generate_from_db('first')
```

Comment:
I am not clear on why the first mlbf would generate differences if we create all the blocks _before_ we generate the bloom filter from the DB.

Reply:
If there is no previous filter to compare to, then the filter will consider everything in the DB a "change" because, well, it is. Nothing to something.

T0: we create all the blocks
T1: we create the first_mlbf (it includes all of the "changes" from T0. since there is no previous filter to compare to everything is a change)
T2: we create and update some records
T3: we create another filter and now track the changes from T1 to T3

Reply:
FWIW that is existing behavior and is how regenerating the filter works.

Reply:
OK so it looks like we don't create all the blocks for all the versions there, which is what I assumed. The first two versions are not blocked at this point. We block them after the initial MLBF is generated.

Reply:
Correct, following the spec in the test description. I think you should now understand [why 1,2 were already unblocked](https://github.com/mozilla/addons-server/pull/22826#discussion_r1837935103). Question is, how could I update the test/comments to make it more clear from the start?